### PR TITLE
Show express payment methods in the Cart block

### LINKED ISSUE #2433: Add Support for Express Payment Buttons in the Cart Block
**Describe the bug**

I found that Stripe is supported as per https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/master/docs/testing/cart-checkout#known-limitations

But seems Stripe's payment request buttons are not supported in the cart block yet? 

https://d.pr/i/pY5Ta2 
Full Size: https://d.pr/i/pY5Ta2

**To Reproduce**

1. Enable Payment Request buttons in Stripe. 
2. Add a product to cart
3. Go to cart page that's using the cart block 
4. You won't see the payment request button

**Expected behavior**
Payment request buttons should show on the cart page.

**WordPress (please complete the following information):**

* Core version: 5.4.1
* WooCommerce version: 4.1.0
* WooCommerce Blocks version: 2.6.0-beta3
* Site language: English
* Any other plugins installed: No

**Desktop (please complete the following information):**

* OS: MacOS Catalina, 10.15.4 
* Browser Chrome - Version 81.0.4044.129 (Official Build) (64-bit)

### PR DESCRIPTION
**Pull request by Aljullu**
Fixes #2433.

This PR:
1. Adds express payment methods to the Cart block.
2. Makes the 'Proceed to Checkout' button in Cart to be sticky _only_ if it's below the viewport. Otherwise it gets positioned normally. That was discussed with @LevinMedia.

### Screenshots

_Partially sticky 'Proceed to Checkout' button:_

_Express payment methods in Cart block:_

### How to test the changes in this Pull Request:

1. Go to the Cart block in a mobile or with a narrow window size and verify the 'Proceed to Checkout' button is no longer sticky once you pass its expected position.
2. If you have express payment methods set up, verify they appear in the Cart block and clicking on them triggers the payment process.

### Release notes

With the introduction of Express payment methods in the Cart block, we updated several CSS class names of the elements responsible to display them:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/theming/class-names-update-330.md

### Changelog

> Express payment methods are now visible in the Cart block too and 'Proceed to Checkout' button stops being sticky after passing it's expected position on mobile.

### CODE REVIEW

File: assets/js/base/components/payment-methods/express-checkout.js
Comment (haszari) [inline assets/js/base/components/payment-methods/express-checkout.js:77]: The breakdown into different components here feels like a patch, rather than logical components. I'm guessing this is complex because we want to share the notices container?

From the high level we have two components we need: 

- `CartExpressPayment` .. or `CartSubmitButtons` (encompassing both express pay and proceed to cart)
- `CheckoutExpressPayment`

The `isCompact` prop and the two variations of `[Compact]ExpressCheckoutContainer` don't feel like a good match for the UI.

I don't think this is a blocker, but if you think we can make these components easier to understand by refactoring, go for it :)

Related – I'm guessing some of this is intended to be used elsewhere, e.g. express payment on product page. Great if we can handle that variation now, but I'd be wary of generalising too early.

**Comment by haszari** [comment]
Actually – one more thing!

When viewing the cart on mobile, should the express pay button be sticky like the `Proceed to checkout` button? This is what I see:

Screen Shot 2020-08-14 at 3 34 05 PM

> Go to the Cart block in a mobile or with a narrow window size and verify the 'Proceed to Checkout' button is no longer sticky once you pass its expected position.

I'm a little confused - is this an essential part of this PR? I could see how it's related and/or convenient to implement here, but could be separate potentially. I might have missed some test cases for this.

File: assets/js/base/hooks/use-position-relative-to-viewport.js
Comment (Aljullu) [reply assets/js/base/hooks/use-position-relative-to-viewport.js]: Thanks for the suggestions, I implemented them in: aec3d94fecec6df5816f47ad267974490797fb16. :slightly_smiling_face:

File: assets/js/blocks/cart-checkout/cart/checkout-button/index.js
Comment (Aljullu) [reply assets/js/blocks/cart-checkout/cart/checkout-button/index.js]: > Do we need any fallback for internet explorer? (probably having it scroll into view is fine..)

I don't think so, the sticky behavior is only available on mobile so no need to add support for IE11.

File: assets/js/base/hooks/use-position-relative-to-viewport.js
Comment (Aljullu) [reply assets/js/base/hooks/use-position-relative-to-viewport.js:43]: Added in 9a3cbba31bfffdca32db74e5fc594540e7b1b79f. :+1:

File: assets/js/base/components/payment-methods/express-checkout.js
Comment (Aljullu) [reply assets/js/base/components/payment-methods/express-checkout.js]: In a818772f9fb970a705e3dc6427c9ca53b93eac6d I split `ExpressCheckoutFormControl` into two components, one to be used in the Cart block and the other in the Checkout block. Let me know how it looks. :slightly_smiling_face:

**Comment by Aljullu** [comment]
Thanks for the review @haszari. I applied most of the feedback in the last commits and answered some of it inline. :point_up_2: 

> Vertical rhythm / padding is off

Fixed in 365d4fdac8538257299b8c6dc1d812cfb3273cf2.

> In the editor, the express pay button doesn't use the Stripe appearance setting (dark/light/outlined) - see screenshot. I think this is shippable but might be worth logging a follow up issue (screenshot below).

I tested and it seems it's affecting Checkout too, can you confirm? If that's the case, let's go for creating an issue (feel free to open it yourself or I will take care of it), so we don't block this PR.

> When viewing the cart on mobile, should the express pay button be sticky like the `Proceed to checkout` button?

It shouldn't, the reason is that if there are several express payment methods it could take too much vertical space. This was initially asked in https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2433#issuecomment-670011261 and later discussed in p1596729028166700-slack-C8X6Q7XQU.

> I'm a little confused - is this an essential part of this PR? I could see how it's related and/or convenient to implement here, but could be delivered separately.

It could be split into its own PR, but I think both features need to be merged at once. Adding Express payment methods to the Cart block without updating the UI would have looked weird (in the Slack chat above there are some screenshots of it).

> I didn't fully understand why IntersectionObserver is needed over a more basic "sticky" implementation

What would be a "more basic sticky"? The other option would have been adding a scroll event listener but for performance reasons I think IntersectionObserver is preferred. And considering there was no need to support IE11 given that it's a mobile specific feature, I thought that was the way to go.

I also considered using an external library, but considering `usePositionRelativeToViewport()` is ~40 lines of code, I thought there was no need to add another dependency.

---

I know this PR has been approved, but I would appreciate another review from you or somebody else from @woocommerce/rubik given that there are some changes/additions in the last commits.

File: assets/js/base/hooks/use-position-relative-to-viewport.js
Comment (haszari) [reply assets/js/base/hooks/use-position-relative-to-viewport.js]: Nice!

File: assets/js/blocks/cart-checkout/cart/checkout-button/index.js
Comment (haszari) [reply assets/js/blocks/cart-checkout/cart/checkout-button/index.js]: Ah - fair point. Though just to confirm, does the sticky behaviour kick in based on window size? If so it can happen on any platform. But I agree, no need for fallback as this is a progressive enhancement.

File: assets/js/base/components/payment-methods/express-checkout.js
Comment (haszari) [reply assets/js/base/components/payment-methods/express-checkout.js]: Looks good to me – naming them based on cart/checkout is much more direct. Can abstract out more later as needed.

I see the class names still use the `compact` naming rather than cart/checkout classes. Is this intentional? I'd assume you'd either style it based on a container class (e.g. `.*cart*`), or the class name would align with component name (e.g. `wc-block-components-express-payment wc-block-components-express-payment__cart` - use BEM modifier for cart vs checkout?)

Up to you how you handle this - happy with whatever you decide.

**Review by haszari — APPROVED** [review]
Looking good @Aljullu - I like the changes, especially the new test and the components focused on UI (cart/checkout). Note one inline comment about classes vs component names - I'd prefer these are consistent.

In today's testing, for some reason Apple Pay isn't working for me. Note that I don't have Apple Pay fully set up - I haven't added a card. So the behaviour may have changed in a recent Safari/OS update. 

So I wasn't able to test the express pay aspect of this. So ideally this would be retested by someone with Apple Pay and Chrome Pay set up and working (I'm guessing you've covered this in your testing). Also I haven't logged the follow up issue for Stripe appearance setting since I can't get a screenshot now - I'll leave that to you.

I did some regression testing of cart & checkout in various browsers and the new sticky behaviour. Didn't see any issues 👍 

> > I didn't fully understand why IntersectionObserver is needed over a more basic "sticky" implementation

> What would be a "more basic sticky"?

Thanks for explaining – I agree that `IntersectionObserver` is a good fit because it's a native API (~ performance). I was thinking of classic `onscroll` approach 😄 

One thing I noticed when testing – the sticky behaviour is set up to work on "mobile" only (i.e. below an arbitrary width breakpoint). I understand this is by design, but I wonder if it's useful and simpler to have it on all the time. Themes and users/browsers (e.g. window size, page layout) vary wildly, so IMO it's always a good thing that the `Proceed to checkout` button is visible. cc @LevinMedia 

For example, in the theme I was testing (Hitchcock by Anders Noren), the button is off-screen for a typical 1-2 item cart, depending on window width:

Screen Shot 2020-08-20 at 11 10 04 AM

File: assets/js/base/hooks/use-position-relative-to-viewport.js
Comment (Aljullu) [reply assets/js/base/hooks/use-position-relative-to-viewport.js:62]: If I'm not wrong, `root` is expected to be the scrollable ancestor of the observed element. In this case, there is no scrollable ancestor other than the browser window, so it should be `null`.

Just to confirm, I tested setting `root: document.body.querySelector( '.entry-content' )` and it didn't seem to work because `isIntersecting` was always `true`.

File: assets/js/blocks/cart-checkout/cart/checkout-button/index.js
Comment (Aljullu) [reply assets/js/blocks/cart-checkout/cart/checkout-button/index.js:70]: That was a bit tricky. If we remove the no-sticky `.wc-block-cart__submit-container` when the sticky version is shown, then `.wc-block-cart__submit` becomes of height=0. In that case, it might become fully visible, so we would remove the sticky container and add again the no sticky one, which might cause `.wc-block-cart__submit` not to be fully visible again. That would cause infinite rerenders.

Hopefully a GIF will explain it better than myself: :slightly_smiling_face: 
Peek 2020-08-20 10-26

To fix that, we need `.wc-block-cart__submit` to always have the same height. I think the cleanest approach to that is to always render the no-sticky container.

I added an explanatory comment in the code in 20b11057e2e9276f25274856df2e223d93760aa1.

File: assets/js/blocks/cart-checkout/cart/checkout-button/style.scss
Comment (Aljullu) [reply assets/js/blocks/cart-checkout/cart/checkout-button/style.scss:32]: I'm not sure to understand what you mean. The CSS breakpoint is there to make the sticky container to be displayed only on small viewports.

File: assets/js/base/hooks/use-position-relative-to-viewport.js
Comment (Aljullu) [reply assets/js/base/hooks/use-position-relative-to-viewport.js]: Good point, I created a shim object in 828d4dd3c5b86c5117be875efe3173de0620cd8a.

File: assets/js/base/hooks/use-position-relative-to-viewport.js
Comment (nerrad) [reply assets/js/base/hooks/use-position-relative-to-viewport.js:63]: 👍 Thanks for the explanation. That works for me.

File: assets/js/blocks/cart-checkout/cart/checkout-button/index.js
Comment (nerrad) [reply assets/js/blocks/cart-checkout/cart/checkout-button/index.js:70]: Works for me 👍 Thanks for the explanation.

File: assets/js/base/utils/get-intersection-observer.js
Comment (nerrad) [inline assets/js/base/utils/get-intersection-observer.js:8]: Maybe add a `@todo` comment here to remove this shim when we drop support for IE11?

File: assets/js/blocks/cart-checkout/cart/checkout-button/style.scss
Comment (nerrad) [reply assets/js/blocks/cart-checkout/cart/checkout-button/style.scss:32]: What I mean is that is that it looks like the `.wc-block-cart__submit-container--sticky` will only be rendered when the `positionRelativetoViewport === 'below'` (as found in [this code block](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/828d4dd3c5b86c5117be875efe3173de0620cd8a/assets/js/blocks/cart-checkout/cart/checkout-button/index.js#L66-L70)). So there's no need for a display property? 

However, looking the code again, it looks like the IntersectionObserver `positionRelativeToViewport` is going to be set regardless of viewport width and thus the need to add `display:none;` here because the sticky container will always be present given the conditions across various viewport widths. I think what tripped me up when doing the review was an [earlier response you gave](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3004#discussion_r473054340) and I was conflating that in my mind that the intersection observation would only be happening in mobile views:

> I don't think so, the sticky behavior is only available on mobile so no need to add support for IE11.

Given the new perspective of this, I think I understand why the css is needed now.

File: assets/js/base/components/payment-methods/express-checkout.js
Comment (Aljullu) [reply assets/js/base/components/payment-methods/express-checkout.js]: Good point. I was trying to avoid breaking old class names, but probably this is a good opportunity to update them and reorganize them a bit. I did it in 88be9aa2757db307ada24d29c0bbb09f2dc16bc4.

**Comment by Aljullu** [comment]
> Also I haven't logged the follow up issue for Stripe appearance setting since I can't get a screenshot now - I'll leave that to you.

Issue created: #3033.

> So I wasn't able to test the express pay aspect of this. Ideally this would be retested by someone with Apple Pay and Chrome Pay set up and working (I'm guessing you've covered this in your testing - if so maybe good to go).

I don't have any Apple device so I could only test Chrome Pay. @nerrad do you have both payment methods set up? It would be good to test that when there is more than one express payment button, they are properly aligned. I already tested it creating mock buttons, but would be good to double-check with real buttons.

> One thing I noticed when testing – the sticky behaviour is set up to work on "mobile" only (i.e. below an arbitrary width breakpoint). I understand this is by design, but I wonder if it's useful and simpler to have it on all the time. Themes and users/browsers (e.g. window size, page layout) vary wildly

Right... currently it depends on CSS media queries instead of our [container queries](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/base/hooks/use-container-queries.js). I think that was so the sticky button is only displayed in small viewports, no matter if the theme is narrow or wide. We can change it in a follow-up if needed.

File: assets/js/base/utils/get-intersection-observer.js
Comment (Aljullu) [reply assets/js/base/utils/get-intersection-observer.js:8]: Done: b2e3d05ec6e9d371d3c8f2068cb86216133b669d. :slightly_smiling_face:

**Pull request merged**
